### PR TITLE
fix: [CDS-70676]: pass correct fieldYaml in case of parallel stages or steps

### PR DESCRIPTION
**Pull request by shalini-agr**
## Describe your changes

**Issue:** When user has an execution input in a step or stage inside parallel, the field yaml returned by BE contains all the steps or stages present in the parallel node.
API endpoint: `GET /pipeline/execution-input/{nodeExecutionId}`
For eg: Suppose we have a stage1 and 2 in parallel with stage1 having executionInput, while on stage1, the user would expect the fieldYaml to be something like: 
```
stage:
    identifier: stage1
    name: ...
```
        
But currently, this is returned
```
stages:
   stage:
       identifier: stage1
       name: ...
   stage: 
       identifier: stage2
       name:...
```
i.e. entire yaml inside parallel node.

**Changes:** Use recursion to get the yaml of the required stage or step in parallel which matches the identifier. As a result, return the correct field yaml of the requested stage or step when present inside the parallel node. 

**Impact:** The field yaml will only contain the requested stage or step yaml when present inside the parallel node.

**Testing:** Tested locally with a pipeline having parallel steps and stages with step group and looping strategy as well.
**Unit Test:** Added a unit test to validate the yaml returned in case of a stage in parallel. 
**Automation:** In this case, automation isn't needed as we only need to validate the returned field yaml which are we are doing via unit test.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49375)

### CODE REVIEW

File: 953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java [hunk 1]
```diff
@@ -485,10 +485,17 @@ public static JsonNode getNodeYaml(String yaml, Ambiance ambiance) {
     } catch (IOException e) {
       throw new InvalidYamlException("Yaml could not be converted to YamlNode. Please check if the yaml is correct.");
     }
-    for (Level level : ambiance.getLevelsList()) {
+    for (int i = 0; i < ambiance.getLevelsList().size(); i++) {
```
Comment (naidu-harness) [inline 953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java:488]: why this change?

Comment (shalini-agr) [reply 953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java:488]: In the case of a parallel node, we need to match the identifier of the next level, so we are using the index of the level to get to the next level to get its identifier.

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49375)* status: 5 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java` line 492 at r3](https://reviewable.io/reviews/harness/harness-core/49375#-NYIvIMXATq_kjcfyAFn:-NYJ0hcSARe9nIHNx_W3:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d210a659adbe8f88d815c110c0eb75c8c9ce8ed3/953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java#L492)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Use YamlFIeldNameConstants
</blockquote></details>

Done.
___
*[`953-yaml-commons/src/test/java/io/harness/pms/yaml/YamlNodeTest.java` line 217 at r2](https://reviewable.io/reviews/harness/harness-core/49375#-NYIH50H8O8zie4nqf71:-NYJ0sgPCZQeIucGG3WM:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fdc61cb401b8c0d6c8eeff7803a890f864bf29c3/953-yaml-commons/src/test/java/io/harness/pms/yaml/YamlNodeTest.java#L217)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

@shalini-agr can you add pipeline yaml is different and its file path
</blockquote></details>

Done.

**Comment by brijeshdhakar** [comment]
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java` line 493 at r3](https://reviewable.io/reviews/harness/harness-core/49375#-NYJ7RcR6rJybsdXdJwt:-NYJ7RcR6rJybsdXdJwu:b-kekh8l) ([raw file](https://github.com/harness/harness-core/blob/d210a659adbe8f88d815c110c0eb75c8c9ce8ed3/953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java#L493)):*
> ```Java
>         for (YamlNode node : currentNode.asArray()) {
>           if (node.gotoPath("parallel") != null) {
>             node = new YamlNode(getNodeYaml(node.gotoPath("parallel"), remainingLevels));
> ```

Discussed offline. We can return the method from here.

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49375)* status: 5 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java` line 488 at r2](https://reviewable.io/reviews/harness/harness-core/49375#gh-1233737539:-NYJAOlW-nze-0BMoDIA:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fdc61cb401b8c0d6c8eeff7803a890f864bf29c3/953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java#L488)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

done
</blockquote></details>

Done.
___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java` line 493 at r3](https://reviewable.io/reviews/harness/harness-core/49375#-NYJ7RcR6rJybsdXdJwt:-NYJAPQEEmQSFJfIXCwF:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d210a659adbe8f88d815c110c0eb75c8c9ce8ed3/953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java#L493)):*
<details><summary><i>Previously, brijeshdhakar (Brijesh Dhakar) wrote…</i></summary><blockquote>

Discussed offline. We can return the method from here.
</blockquote></details>

Done.

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 4 files at r3, 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49375)* status: 5 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Comment by shalini-agr** [comment]
trigger uts

**Comment by shalini-agr** [comment]
trigger uts

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 1 of 2 files at r4, 1 of 1 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49375)* status: 5 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 1 of 1 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49375)* status: 1. unresolved discussions.

**Comment by archit-harness** [comment]
trigger review

**Comment by sahilhindwani-harness** [comment]
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java` line 481 at r7](https://reviewable.io/reviews/harness/harness-core/49375#-NYMbX2aC9q85gcO3QSm:-NYMbX2aC9q85gcO3QSn:b-jrgzx7) ([raw file](https://github.com/harness/harness-core/blob/c54795f57e0a5f7c30bb239665fa9e8f202f97a1/953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlNode.java#L481)):*
> ```Java
>   // get the field/node yaml from the complete pipeline yaml.
>   public static JsonNode getNodeYaml(YamlNode yamlNode, List<Level> levelsList) {
>     YamlNode currentNode = yamlNode.gotoPath("");
> ```

Can we add a comment as to why we are doing this?

**Review by sahilhindwani-harness — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 1 of 1 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49375)* status: 1. unresolved discussions.
